### ISSUE TITLE
Error when addon trying to access/dev/i2c-1

### ISSUE THREAD

**Issue by awiamo** [hassio-addons/addon-mqtt-io#35]
# Problem/Motivation
I have RPi-OS and hassio supervised installed on my RPi4. When I try to run addon-mqtt-io with a BME280 sensor connected locally via GPIO I2C, I get below error:

> PermissionError: [Errno 1] Operation not permitted: '/dev/i2c-1'

The same problem occurs on RPi3b running on the latest haos-9.3 system. I tried to diagnose the problem, the conclusions are as below.

## Actual behavior

Addon Logs:
```bash
  File "/usr/lib/python3.10/site-packages/mqtt_io/server.py", line 128, in _init_module
    return module_class(module_config)
  File "/usr/lib/python3.10/site-packages/mqtt_io/modules/sensor/__init__.py", line 22, in __init__
    self.setup_module()
  File "/usr/lib/python3.10/site-packages/mqtt_io/modules/sensor/bme280.py", line 38, in setup_module
    self.bus = SMBus(self.config["i2c_bus_num"])
  File "/usr/lib/python3.10/site-packages/smbus2/smbus2.py", line 280, in __init__
    self.open(bus)
  File "/usr/lib/python3.10/site-packages/smbus2/smbus2.py", line 310, in open
    self.fd = os.open(filepath, os.O_RDWR)
PermissionError: [Errno 1] Operation not permitted: '/dev/i2c-1'
[11:29:09] WARNING: MQTT IO crashed, halting add-on
s6-rc: info: service legacy-services: stopping
[11:29:10] INFO: MQTT IO stopped
s6-rc: info: service legacy-services successfully stopped
s6-rc: info: service legacy-cont-init: stopping
s6-rc: info: service legacy-cont-init successfully stopped
s6-rc: info: service fix-attrs: stopping
s6-rc: info: service fix-attrs successfully stopped
s6-rc: info: service s6rc-oneshot-runner: stopping
s6-rc: info: service s6rc-oneshot-runner successfully stopped
```
[mqtt-io_log.txt](https://github.com/hassio-addons/addon-mqtt-io/files/9952633/mqtt-io_log.txt)

## Steps to reproduce

The plugin configuration is as follows  **/config/mqtt-io/config.yml**:
```bash
mqtt:
  host: core-mosquitto
  user: homeassistant
  password: [hidden]
  topic_prefix: mqtt-io
sensor_modules:
  # BME280 on I2C bus
  - name: bme_sensor
    module: bme280
    i2c_bus_num: 1
    chip_addr: 0x76
sensor_inputs:
  - name: sen1_temperature
    module: bme_sensor
```

Sensor BME280 connected to i2c-1 interface GPIO RPi4.

### 1. Read GPIO i2c from RPI4 host:
Check the operation of the BME280 sensor directly from the host:
```bash
[root@home /]# i2cdetect -y 1
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:                         -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: -- -- -- -- -- -- 76 --
```
```bash
[root@home /home/awitt/temp]# read_bme280 --i2c-address 0x76
1002.46 hPa
  49.42 %
  24.11 C

```
### 2. Read GPIO i2c from container addon-mqtt-io
Then I run the same commands in the mqtt-io container running in HA. I'm starting the add-on with minimal configuration **/config/mqtt-io/config.yml**

```bash
mqtt:
  host: core-mosquitto
  user: homeassistant
  password: [hidden]
  topic_prefix: mqtt-io
```
then from the host CLI enter the mqtt-io container:
```bash
[root@home /]# docker exec -it addon_a0d7b954_mqtt-io bash
```
then in the container:
```bash
apk add --no-cache i2c-tools && i2cdetect -y 1
```
as a result of the command, I get the response:
```bash
a0d7b954-mqtt-io:~# apk add --no-cache i2c-tools && i2cdetect -y 1
fetch https://dl-cdn.alpinelinux.org/alpine/v3.16/main/aarch64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.16/community/aarch64/APKINDEX.tar.gz
(1/1) Installing i2c-tools (4.3-r1)
Executing busybox-1.35.0-r17.trigger
OK: 89 MiB in 54 packages
Error: Could not open file `/dev/i2c-1': Operation not permitted
a0d7b954-mqtt-io:~#
```
[mqtt-io_docker_inspect.txt](https://github.com/hassio-addons/addon-mqtt-io/files/9952643/mqtt-io_docker_inspect.txt)

### 3. Read GPIO i2c from new alpine container
I'm running the alpine: latest container with permissions:
SYS_RAWIO
/dev/mem
/dev/gpiomem

```bash
[root@home /]# docker run -it --rm --cap-add SYS_RAWIO --device /dev/mem --device /dev/gpiomem alpine:latest sh
```
and run i2c test:
```bash
[root@home /]# docker run -it --rm --cap-add SYS_RAWIO --device /dev/mem --device /dev/gpiomem alpine:latest sh
/ # apk add --no-cache i2c-tools && i2cdetect -y 1
fetch https://dl-cdn.alpinelinux.org/alpine/v3.16/main/aarch64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.16/community/aarch64/APKINDEX.tar.gz
(1/1) Installing i2c-tools (4.3-r1)
Executing busybox-1.35.0-r17.trigger
OK: 6 MiB in 15 packages
Error: Could not open file `/dev/i2c-1' or `/dev/i2c/1': No such file or directory
/ #
```

### 4. Read GPIO i2c from new alpine container (/dev/i2c-1)
I'm running the alpine: latest container with permissions:
/dev/i2c-1

```bash
[root@home /]#  docker run -it --rm --device /dev/i2c-1 alpine:latest sh
/ # apk add --no-cache i2c-tools && i2cdetect -y 1
fetch https://dl-cdn.alpinelinux.org/alpine/v3.16/main/aarch64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.16/community/aarch64/APKINDEX.tar.gz
(1/1) Installing i2c-tools (4.3-r1)
Executing busybox-1.35.0-r17.trigger
OK: 6 MiB in 15 packages
     0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
00:                         -- -- -- -- -- -- -- --
10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
40: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
70: -- -- -- -- -- -- 76 --
/ #

```
## System Info
```bash
[root@home /]# uname -a
Linux home 5.15.74-v8+ #1595 SMP PREEMPT Wed Oct 26 11:07:24 BST 2022 aarch64 GNU/Linux

[root@home /]# cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```
 Add-on version: 0.1.3
 You are running the latest version of this add-on.
 System: Debian GNU/Linux 11 (bullseye)  (aarch64 / raspberrypi4-64)
 Home Assistant Core: 2022.10.5
 Home Assistant Supervisor: 2022.10.2

## Questions
1. How do I get the plug to work? In the docker container called with the parameters `--cap-add SYS_RAWIO --device / dev / mem --device / dev / gpiomem` it wont work.

2. Changing the parameters `SUBSYSTEM=="i2c-dev", GROUP="i2c", MODE="0666"` oraz `SUBSYSTEM=="gpio", GROUP="gpio", MODE="0666"` it doesn't help

3. The problem seems to be resolved after mapping `/dev/i2c-1`. So, should this parameter be added to the plugin configuration?

**Comment by awiamo** [comment]
### Update
In the file [config.yaml](https://github.com/hassio-addons/addon-mqtt-io/blob/main/mqtt-io/config.yaml) is a configuration::

```bash
    devices:
    - /dev/mem
    - /dev/gpiomem
```
I would expect a running docker addon containing the devices listed in the config. However, a dump of a running container via **docker inspect addon_a0d7b954_mqtt-io** shows that docker is started without mapping these devices.

```bash
"Devices": null,
```

###Steps to reproduce
Get the file [run.tpl](https://gist.github.com/efrecon/8ce9c75d518b6eb863f667442d7bc679 "run.tpl")  and and manually start the docker container from the CLI:

```bash
docker inspect --format "$(<run.tpl)" name_or_id_of_running_container
```
this will generate the configuration to run. If I add the command **--devices /dev/i2c-1** to mannualy generated config it starts working properly.

**Comment by thecode** [comment]
> this will generate the configuration to run. If I add the command **--devices /dev/i2c-1** to mannualy generated config it starts working properly.

@awiamo would you like to create a PR to add this to the config?

**Comment by awiamo** [comment]
Even if you add /dev/i2c-1 to add-on configuration it won't work. Since the docker container is started with empty "Devices: null" - this is probably the source of the problem.
@thecode I don't need to do PR. I could fork this repository and change what I want - but I already know that it won't work because the problem may be in parsing the configuration by hassio.

**Comment by mousetwentytwo** [comment]
Can confirm this issue, seems like the container started by ha does not have the device accesses added.
Do you know of a non temporary workaround? 
I mean one that keeps working after system restarts?

**Comment by awiamo** [comment]
>Do you know of a non temporary workaround?
>I mean one that keeps working after system restarts?
As I wrote before: the only "elegant" solution would be to create a repository fork and add the device /dev/i2c-xx appropriate for your configuration in the plugin.  You can modify the plugin's configuration in the `addons.json` file (I see that it does not overwrite the settings after reboot), but this change will probably not be permanent (e.g. in the case of plugin update).

**Comment by awiamo** [comment]
Regarding the reported problem:

In the documentation [mqtt-io.app](https://mqtt-io.app/2.2.7/#/deployment/docker) referring to running in a docker container, it clearly indicated (in the last two paragraphs) that when using i2c devices, it is required to run the docker image with these devices.

So,
- the plugin will work properly if we want to read the status of individual GPIO pins because the code uses device `/dev/gpiomem` in this case.

```mqtt:
  host: core-mosquitto
  user: homeassistant
  password: [hidden]
  topic_prefix: mqtt-io

# GPIO
gpio_modules:
  # Use the Raspberry Pi built-in GPIO
  - name: rpi
    module: raspberrypi

digital_outputs:
  # Pin 1 is an output connected to a light
  - name: porch_light
    module: rpi
    pin: 1
```
- Lack of permissions in the docker container to the correct device `/dev/i2c-xx` will result in an error when it starts. Because the code uses the smbus library and tries to connect to the `/dev/i2c-xx` device, not the `/dev/gpiomem` device as in the previous example. The code below will not work correctly.

```
mqtt:
  host: core-mosquitto
  user: homeassistant
  password: [hidden]
  topic_prefix: mqtt-io
  
# Sensors
sensor_modules:
  # An LM75 sensor attached to the I2C bus
  - name: lm75_sensor
    module: lm75
    i2c_bus_num: 1
    chip_addr: 0x48

sensor_inputs:
  # The configuration of the specific sensor value to use (LM75 only has temperature)
  - name: porch_temperature
    module: lm75_sensor
```

## Proposed changes

I would suggest adding devices `/dev/i2c-xx` in the plugin. However, it would be best to define devices in the GUI in Hassio plugin because different boards have a different number of i2c ports.

**Comment by frenck** [comment]
> @thecode I don't need to do PR. I could fork this repository and change what I want - but I already know that it won't work because the problem may be in parsing the configuration by hassio.

So, this is kinda demotivating for any maintainer/contributor. I don't mind someone setting such a tone at me, but not to others, and definitely not when addressing people directly using an GitHub mention. I suggest adjusting your tone, providing your sincere apologies, or instead, just shut the hell up, fork it, and get lost.

Sorry for the strong language.

../Frenck

**Comment by frenck** [comment]
PR #38 enables the maximum access an add-on can possibly get to devices; That is how far an add-on can possibly go.

**Comment by awiamo** [comment]
Sorry if I offended you with this post. I myself spent several hours analyzing the problem and describing it in the best possible way. The fact that I pointed out the same problem twice was taken as "creating a PR" to force a change. The response to my analysis seemed a bit ignorant. Look at it from my side. Anyway, I still tried to explain the problem and describe it so that it was 100% sure that it is not the fault of a bad configuration.

Sorry again. It wasn't my intention to offend you. You guys are doing a really good job and I really appreciate it.
